src/client/opamTreeCommand.ml

```diff
@@ -0,0 +1,373 @@
+  OpamSolver.PkgGraph.out_degree graph p = 0
+
+let cut_leaves (mode: [ `succ | `pred]) ~names ~root st graph =
+  let open OpamSolver.PkgGraph in
+  let fold, is_final =
+    match mode with
+    | `succ -> fold_succ, is_leaf graph
+    | `pred -> fold_pred, is_root graph
+  in
+  (* compute the packages which are connected to one of the `names` *)
+  let rec go package set =
+    fold (fun p ps ->
+      if OpamPackage.Set.mem p ps then ps
+      else
+        let ps = OpamPackage.Set.add p ps in
+        if not (is_final p) then go p ps
+        else ps
+    ) graph package set
+  in
+  let packages = names |> OpamPackage.Set.fold go names in
+  (* cut leaves not belonging to the packages *)
+  OpamPackage.Set.diff st.installed packages
+  |> OpamPackage.Set.iter (OpamSolver.PkgGraph.remove_vertex graph);
+  (* return the new roots and the new graph *)
+  OpamPackage.Set.inter root packages, graph
+
+let build_deps_forest st universe ?(post=false) ?(dev=false) ?(doc=false) ?(test=false)
+    ?(tools=false) filter names =
+  let names = installed st names in
+  let root, graph =
+    let graph =
+      OpamSolver.dependency_graph
+        ~depopts:false ~build:false ~post ~installed:true
+        universe
+    in
+    let root =
+      st.installed |> OpamPackage.Set.filter (is_root graph)
+    in
+    match OpamPackage.Set.is_empty names, filter with
+    | false, Roots_from  -> names, graph
+    | false, Leads_to -> cut_leaves `pred ~names ~root st graph
+    | true, _ -> root, graph
+  in
+  let condition_map = build_condition_map ~post ~doc ~test ~tools ~dev st in
+  let rec build visited package node =
+    if visited |> OpamPackage.Set.mem package then
+      let node =
+        match node with
+        | Root p -> Root p (* but impossible *)
+        | Dependency x -> Dependency { x with is_dup = true }
+      in
+      visited, Tree.create node
+    else
+      let visited = visited |> OpamPackage.Set.add package in
+      let conditions = condition_map |> OpamPackage.Map.find package in
+      let succ = OpamSolver.PkgGraph.succ graph package in
+      let visited, children =
+        OpamStd.List.fold_left_map (fun visited package ->
+          let satisfies = conditions |> OpamPackage.(Name.Map.find_opt package.name) in
+          let child_node = Dependency { package; satisfies; is_dup = false } in
+          build visited package child_node
+        ) visited succ
+      in
+      visited, Tree.create ~children node
+  in
+  let build_root visited package =
+    build visited package (Root package)
+  in
+  root
+  |> OpamPackage.Set.elements
+  |> OpamStd.List.fold_left_map build_root OpamPackage.Set.empty
+  |> snd
+
+let build_revdeps_forest st universe ?(post=false) ?(dev=false) ?(doc=false)
+    ?(test=false) ?(tools=false) filter names =
+  let names = installed st names in
+  let root, graph =
+    let graph =
+      OpamSolver.dependency_graph
+        ~depopts:false ~build:false ~post ~installed:true
```

Comment:
Because we currently don't have an option to show build dependencies.

I think we can add the option by editing here and `build_condition_map`. Should we do that?

Reply:
My main concern was that users may struggle to distinguish build depenedencies from the normal ones. But I saw your branch and it shows as:
```
c.1
|-- a.1 [*]
|-- b.1 [*]
'-- c-build.1 (build)
```

So I now think it is ok to just change `false` to `true`.